tools/ci/pr_preview.py

```diff
@@ -0,0 +1,403 @@
+
+        subprocess.check_call([
+            'git',
+            '-c',
+            'credential.username={}"'.format(self._token),
+            '-c',
+            'core.askPass=true',
+            'push',
+            'origin',
+            '--delete',
+            full_ref
+        ])
+
+def is_open(pull_request):
+    return not pull_request['closed_at']
+
+def has_label(pull_request):
+    for label in pull_request['labels']:
+        if label['name'] == LABEL:
+            return True
+
+    return False
+
+def should_be_mirrored(pull_request):
+    return (
+        is_open(pull_request) and
+        pull_request['user']['login'] not in AUTOMATION_GITHUB_USERS and (
+            pull_request['author_association'] in TRUSTED_AUTHOR_ASSOCIATIONS or
+            has_label(pull_request)
+        )
+    )
+
+def is_deployed(host, deployment):
+    response = requests.get(
+        '{}/.git/worktrees/{}/HEAD'.format(host, deployment['environment'])
+    )
+
+    if response.status_code != 200:
+        return False
+
+    return response.text.strip() == deployment['sha']
+
+def synchronize(host, github_project, window):
+    '''Inspect all Pull Requests which have been modified in a given window of
+    time. Add or remove the "preview" label and update or delete the relevant
+    git refs according to the status of each Pull Request.'''
+
+    project = Project(host, github_project)
+    remote = Remote(github_project)
+
+    pull_requests = project.get_pull_requests(
+        time.gmtime(time.time() - window)
+    )
+
+    for pull_request in pull_requests:
+        logger.info('Processing Pull Request #%(number)d', pull_request)
+
+        refspec_trusted = 'prs-trusted-for-preview/{number}'.format(
+            **pull_request
+        )
+        refspec_open = 'prs-open/{number}'.format(**pull_request)
+        revision_latest = remote.get_revision(
+            'pull/{number}/head'.format(**pull_request)
+        )
+        revision_trusted = remote.get_revision(refspec_trusted)
+        revision_open = remote.get_revision(refspec_open)
+
+        if should_be_mirrored(pull_request):
+            logger.info('Pull Request should be mirrored')
+
+            if revision_trusted is None:
+                project.create_ref(refspec_trusted, revision_latest)
+            elif revision_trusted != revision_latest:
+                project.update_ref(refspec_trusted, revision_latest)
+
+            if revision_open is None:
+                project.create_ref(refspec_open, revision_latest)
+            elif revision_open != revision_latest:
+                project.update_ref(refspec_open, revision_latest)
+
+            if project.get_deployment(revision_latest) is None:
+                project.create_deployment(
+                    pull_request, revision_latest
+                )
+        else:
+            logger.info('Pull Request should not be mirrored')
+
+            if not has_label(pull_request) and revision_trusted is not None:
```

Comment:
Does this mean that if a PR with the 'safe for preview' label is mirrored (creating a prs-trusted-for-preview/XYZ ref), that ref will never be removed even after the PR is closed? If that's intended, I think maybe I don't understand the flow here.

Reply:
That's correct; trust is independent from the open/closed state of the pull request. [The mirroring system only publishes previews for pull requests that have both refs](https://github.com/bocoup/web-platform-tests.live/blob/95770444285e86ec5851f30a8a5395ed1d79216b/src/mirror-pull-requests.sh#L41), so removing either will cancel the preview.